### PR TITLE
[SPARK-25119][Web UI] stages in wrong order within job page DAG chart

### PR DESCRIPTION
**Pull request by yunjzhang**
if spark job contains multiple tasks , the order in DAG chart might be incorrect.
sample and screen snapshot can be found in jira ticket https://issues.apache.org/jira/browse/SPARK-25119

to fix this issue, just sort RDDOperationGraph Array on "task id" before UIUtils.showDagVizForJob()

## How was this patch tested?
packaging with all existing UT passed. 
run complex query in Spark-sql console with the new jars
check the DAG chart in job page.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala [hunk 1]
```diff
@@ -18,18 +18,18 @@
 package org.apache.spark.ui.jobs
 
 import java.util.Locale
+
 import javax.servlet.http.HttpServletRequest
 
 import scala.collection.mutable.{Buffer, ListBuffer}
 import scala.xml.{Node, NodeSeq, Unparsed, Utility}
-
```
Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: revert this changes in import.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala [hunk 2]
```diff
@@ -337,7 +337,9 @@ private[ui] class JobPage(parent: JobsTab, store: AppStatusStore) extends WebUIP
       store.executorList(false), appStartTime)
 
     val operationGraphContent = store.asOption(store.operationGraphForJob(jobId)) match {
-      case Some(operationGraph) => UIUtils.showDagVizForJob(jobId, operationGraph)
+      case Some(operationGraph) => UIUtils.showDagVizForJob(jobId, operationGraph.sortWith(
+        _.rootCluster.id.replaceAll(RDDOperationGraph.STAGE_CLUSTER_PREFIX, "").toInt
```
Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: Add `getStageId` function in `RDDOperationGraph` to do this will be better.

**Comment by xuanyuanking** [comment]
Please change title to "[SPARK-25199][Web UI] XXX " as we described in http://spark.apache.org/contributing.html. 
```
check the DAG chart in job page.
```
Could you also put the DAG chart screenshot after your fix?

**Comment by yunjzhang** [comment]
thanks for the suggestion, just rename the PR.
before fix
image

after fix
image

**Comment by MaxGekk** [comment]
Probably you put wrong JIRA in the title. `SPARK-25199`  -> `SPARK-25119`

**Comment by yunjzhang** [comment]
my fault, just fixed.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala (hunk 1 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: +1

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: +1. `replaceAll` also seems like the wrong API to use, `substring` seems more correct.

**Comment by vanzin** [comment]
You're sorting on stage ID but your description says task ID. Just a small inconsistency.

This is also against branch-2.3, it should be against master.

**Comment by wangyum** [comment]
How about backport https://github.com/apache/spark/pull/21680?

**Comment by vanzin** [comment]
That looks better indeed. I'll backport that one to 2.3.

@yunjzhang could you close this PR?

**Comment by yunjzhang** [comment]
I respect your decision.
this PR does not fix the order issue from the root.
I saw differences when jobs/tasks were shown in spark 2.1/2.3.
same insert...select was treated as 2 jobs in 2.1, while which was 1 job in 2.3.
If 2.3 can push tasks into an array in right order, extract sort is not necessary in this PR and also PR#21680

**Comment by vanzin** [comment]
If you think there is still an issue, open a PR against master (not 2.3).
